### DISCUSSION TITLE
Single quotes for attributes

### DISCUSSION THREAD

**Discussion by lllama** [a-h/templ#328] [Q&A]
Hello,

Is there a way to have tag attributes wrapped in single quotes rather than double? I'm trying to use templ with htmx and need single quotes to use with the hx-vals attribute. 

(I also need to have dynamic values in the attribute from an array I'm looping over - I think I'd need a componentFunc for that?)

**Comment by a-h** [comment]
Hi, I'm struggling to understand what you mean here.

You can use single quotes for attrbutes in templ, it was tracked in https://github.com/a-h/templ/issues/116 if that's what you mean?

A code example of what you'd like to write in templ, and the HTML output you'd like to see would be helpful.

**Reply by lllama** [reply]
Sorry I think this may be an editor or LSP issue. When I save then things are reformatted to double quotes. I'll double check what's doing that.

**Reply by lllama** [reply]
This is what I'm trying to achieve btw:

```go
package templates

import (
        "github.com/kevinburke/ssh_config"
)

templ Hosts(hosts []*ssh_config.Host) {
    <ul>
        for _, host := range hosts {
            for _, pattern := range host.Patterns {
               if pattern.String() != "*" {
                   <li wails-call="ConnectToHost" hx-vals='{"host": { pattern.String() } }'>{ pattern.String() }</li>
               }
            }
        }
    </ul>
}
```

My other issue is that the `{ pattern.String() }` inside the attribute isn't being evaluated. But I shall explore the docs etc for a solution for that one.

**Reply by lllama** [reply]
Ah - so it seems that if I put single quotes in then `templ generate` doesn't complain but I still get double quotes written in the output:

```go
for _, pattern := range host.Patterns {
    if pattern.String() != "*" {
        _, templ_7745c5c3_Err = templ_7745c5c3_Buffer.WriteString("<li wails-call=\"ConnectToHost\" hx-vals=\"{&#34;host&#34;: { pattern.String() } }\">")

```

i.e. singles are converted to double.

**Reply by a-h** [reply]
Ah, I think you want something more like this:

```go
package main

import "encoding/json"
import "os"

type Host struct {
	Patterns []Pattern
}

func NewPattern(p string) Pattern {
	return Pattern{p: p}
}

type Pattern struct {
	p string
}

func (p Pattern) String() string {
	return p.p
}

func hostMapJSON(p Pattern) string {
	m := map[string]string{
		"host": p.String(),
	}
	v, _ := json.Marshal(m)
	return string(v)
}

templ Hosts(hosts []Host) {
	<ul>
		for _, host := range hosts {
			for _, pattern := range host.Patterns {
				if pattern.String() != "*" {
					<li wails-call="ConnectToHost" hx-vals={ hostMapJSON(pattern) }>{ pattern.String() }</li>
				}
			}
		}
	</ul>
}

func main() {
	hosts := []Host{
		{
			Patterns: []Pattern{
				NewPattern("*"),
				NewPattern("other"),
			},
		},
	}
	Hosts(hosts).Render(context.Background(), os.Stdout)
}
```

The output of this is:

```html
<ul><li wails-call="ConnectToHost" hx-vals="{&#34;host&#34;:&#34;other&#34;}">other</li></ul>
```

It is completely fine for the `"` in the JSON to have become `&#34;`, that's the correct attribute escaping for double quotes.

**Comment by a-h** [comment]
Just following up on this one. I think there was an issue, which was fixed in this commit https://github.com/a-h/templ/commit/668a888d242bd41efde6ad4d164d13dcdcc20870 via this PR https://github.com/a-h/templ/pull/384

When `templ fmt` was ran, double quotes were being converted into single quotes for constant attributes, causing errors.

I'm going to merge a couple more PRs and cut a new release.
